README
Dot -- July, 2016

Simple dot language parser implemented using regular expressions


parser/ChangeLog
2016-07-06  Carlos Linares Lopez  <carlos.linares@uc3m.es>

	* parser: initial version.



src/DOTdefs.h
/* 
  DOTdefs.h
  Description: basic definitions
  ----------------------------------------------------------------------------- 

  Started on  <Fri Jul  1 09:11:32 2016 Carlos Linares Lopez>
  Last update <jueves, 21 julio 2016 10:27:21 Carlos Linares Lopez (clinares)>
  -----------------------------------------------------------------------------

  $Id::                                                                      $
  $Date::                                                                    $
  $Revision::                                                                $
  -----------------------------------------------------------------------------

  Made by Carlos Linares Lopez
  Login   <clinares@atlas>
*/

#ifndef   	DEFS_H_
# define   	DEFS_H_

// regexps used to parse the input files using a subset of the dot specification
// language. For a full description of the dot language, see:
//
//    http://www.graphviz.org/doc/info/lang.html

#include <iostream>
#include <string>

// Specification part
#define GRAPH_TYPE "^[[:space:]]*([[:alnum:]]+)"
#define GRAPH_NAME "^[[:space:]]*([[:alnum:]]*)"
#define BLOCK_BEGIN "^[[:space:]]*\\{"
#define BLOCK_END "^[[:space:]]*}"
#define VERTEX_NAME "^[[:space:]]*([[:alnum:]]+)"
#define EDGE_TYPE "^[[:space:]]*(--|->)"
#define END_OF_STATEMENT "^[[:space:]]*;"
// #define C_COMMENT R"(^[[:space:]]*/*(.|[\r\n])*?*/)"
#define CPP_COMMENT "^[[:space:]]*//.*\n"

#define ATTRIBUTE_BEGIN "^[[:space:]]*\\["
#define ATTRIBUTE_NAME "^[[:space:]]*(([[:alnum:]]|_)+)[[:space:]]*="
#define ATTRIBUTE_VALUE_NEXT "^[[:space:]]*((([+-]?(\\.[[:digit:]]|[[:digit:]]+(\\.[[:digit:]]*)?))|([[:alnum:]])|(_|\".*\")+))[[:space:]]*,"
// #define ATTRIBUTE_VALUE_END "^[[:space:]]*((([+-]?(\\.[[:digit:]]|[[:digit:]]+(\\.[[:digit:]]*)?))|([[:alnum:]])|(_|\".*\")+))[[:space:]]*\\]"
#define ATTRIBUTE_VALUE_END "^[[:space:]]*((([+-]?(\\.[[:digit:]]|[[:digit:]]+(\\.[[:digit:]]*)?))|([[:alnum:]])|(_|\"[^\"]*\")+))[[:space:]]*\\]"
#define ATTRIBUTE_END "^[[:space:]]*\\]"

namespace dot {
  
  using namespace std;

  // the following is a shortcut for writing data on cout
  template<class T> void show_value (const string& line, const T& value, bool verbose)
  {
    if (verbose)
      cout << " [" << line << " <" << value << ">]" << endl;
  }
} // namespace dot

#endif 	    /* !DEFS_H_ */




/* Local Variables: */
/* mode:c++ */
/* fill-column:80 */
/* End: */


parser/system.h
/* system-dependent definitions for parser

   Written for GNU fileutils.
   Adapted for autoproject by James R. Van Zandt.

   Copyright (C) 89, 91, 92, 93, 94, 1996 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

/* Include sys/types.h before this file.  */

#include <sys/stat.h>

#ifdef STAT_MACROS_BROKEN
#undef S_ISBLK
#undef S_ISCHR
#undef S_ISDIR
#undef S_ISFIFO
#undef S_ISLNK
#undef S_ISMPB
#undef S_ISMPC
#undef S_ISNWK
#undef S_ISREG
#undef S_ISSOCK
#endif /* STAT_MACROS_BROKEN.  */

#ifndef S_IFMT
#define S_IFMT 0170000
#endif
#if !defined(S_ISBLK) && defined(S_IFBLK)
#define S_ISBLK(m) (((m) & S_IFMT) == S_IFBLK)
#endif
#if !defined(S_ISCHR) && defined(S_IFCHR)
#define S_ISCHR(m) (((m) & S_IFMT) == S_IFCHR)
#endif
#if !defined(S_ISDIR) && defined(S_IFDIR)
#define S_ISDIR(m) (((m) & S_IFMT) == S_IFDIR)
#endif
#if !defined(S_ISREG) && defined(S_IFREG)
#define S_ISREG(m) (((m) & S_IFMT) == S_IFREG)
#endif
#if !defined(S_ISFIFO) && defined(S_IFIFO)
#define S_ISFIFO(m) (((m) & S_IFMT) == S_IFIFO)
#endif
#if !defined(S_ISLNK) && defined(S_IFLNK)
#define S_ISLNK(m) (((m) & S_IFMT) == S_IFLNK)
#endif
#if !defined(S_ISSOCK) && defined(S_IFSOCK)
#define S_ISSOCK(m) (((m) & S_IFMT) == S_IFSOCK)
#endif
#if !defined(S_ISMPB) && defined(S_IFMPB) /* V7 */
#define S_ISMPB(m) (((m) & S_IFMT) == S_IFMPB)
#define S_ISMPC(m) (((m) & S_IFMT) == S_IFMPC)
#endif
#if !defined(S_ISNWK) && defined(S_IFNWK) /* HP/UX */
#define S_ISNWK(m) (((m) & S_IFMT) == S_IFNWK)
#endif

#ifndef S_IEXEC
#define S_IEXEC S_IXUSR
#endif

#ifndef S_IXUSR
#define S_IXUSR S_IEXEC
#endif
#ifndef S_IXGRP
#define S_IXGRP (S_IEXEC >> 3)
#endif
#ifndef S_IXOTH
#define S_IXOTH (S_IEXEC >> 6)
#endif
#ifndef S_IXUGO
#define S_IXUGO (S_IXUSR | S_IXGRP | S_IXOTH)
#endif

#if !defined(HAVE_MKFIFO)
#define mkfifo(path, mode) (mknod ((path), (mode) | S_IFIFO, 0))
#endif

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

/* <unistd.h> should be included before any preprocessor test
   of _POSIX_VERSION.  */
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifndef STDIN_FILENO
#define STDIN_FILENO 0
#endif

#ifndef STDOUT_FILENO
#define STDOUT_FILENO 1
#endif

#ifndef STDERR_FILENO
#define STDERR_FILENO 2
#endif

/* FIXME: Don't use _POSIX_VERSION.  */
#ifndef _POSIX_VERSION
off_t lseek ();
#endif

#ifdef TM_IN_SYS_TIME
#include <sys/time.h>
#else
#include <time.h>
#endif

/* Since major is a function on SVR4, we can't use `ifndef major'.  */
#ifdef MAJOR_IN_MKDEV
#include <sys/mkdev.h>
#define HAVE_MAJOR
#endif
#ifdef MAJOR_IN_SYSMACROS
#include <sys/sysmacros.h>
#define HAVE_MAJOR
#endif
#ifdef major			/* Might be defined in sys/types.h.  */
#define HAVE_MAJOR
#endif

#ifndef HAVE_MAJOR
#define major(dev)  (((dev) >> 8) & 0xff)
#define minor(dev)  ((dev) & 0xff)
#define makedev(maj, min)  (((maj) << 8) | (min))
#endif
#undef HAVE_MAJOR

#ifdef HAVE_UTIME_H
#include <utime.h>
#endif

/* Some systems (even some that do have <utime.h>) don't declare this
   structure anywhere.  */
#ifndef HAVE_STRUCT_UTIMBUF
struct utimbuf
{
  long actime;
  long modtime;
};
#endif

/* Don't use bcopy!  Use memmove if source and destination may overlap,
   memcpy otherwise.  */

#ifdef HAVE_STRING_H
# if !STDC_HEADERS && HAVE_MEMORY_H
#  include <memory.h>
# endif
# include <string.h>
#else
# include <strings.h>
char *memchr ();
#endif

#include <errno.h>
#ifndef errno
extern int errno;
#endif

#ifdef STDC_HEADERS
#define getopt system_getopt
#include <stdlib.h>
#undef getopt
#else /* not STDC_HEADERS */
char *getenv ();
#endif /* STDC_HEADERS */

/* The following test is to work around the gross typo in
   systems like Sony NEWS-OS Release 4.0C, whereby EXIT_FAILURE
   is defined to 0, not 1.  */
#if !EXIT_FAILURE
# undef EXIT_FAILURE
# define EXIT_FAILURE 1
#endif

#ifndef EXIT_SUCCESS
# define EXIT_SUCCESS 0
#endif

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#else
#include <sys/file.h>
#endif

#ifndef SEEK_SET
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2
#endif
#ifndef F_OK
#define F_OK 0
#define X_OK 1
#define W_OK 2
#define R_OK 4
#endif

#ifdef HAVE_DIRENT_H
# include <dirent.h>
# define NLENGTH(direct) (strlen((direct)->d_name))
#else /* not HAVE_DIRENT_H */
# define dirent direct
# define NLENGTH(direct) ((direct)->d_namlen)
# ifdef HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif /* HAVE_SYS_NDIR_H */
# ifdef HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif /* HAVE_SYS_DIR_H */
# ifdef HAVE_NDIR_H
#  include <ndir.h>
# endif /* HAVE_NDIR_H */
#endif /* HAVE_DIRENT_H */

#ifdef CLOSEDIR_VOID
/* Fake a return value. */
#define CLOSEDIR(d) (closedir (d), 0)
#else
#define CLOSEDIR(d) closedir (d)
#endif

/* Get or fake the disk device blocksize.
   Usually defined by sys/param.h (if at all).  */
#ifndef DEV_BSIZE
#ifdef BSIZE
#define DEV_BSIZE BSIZE
#else /* !BSIZE */
#define DEV_BSIZE 4096
#endif /* !BSIZE */
#endif /* !DEV_BSIZE */

/* Extract or fake data from a `struct stat'.
   ST_BLKSIZE: Optimal I/O blocksize for the file, in bytes.
   ST_NBLOCKS: Number of 512-byte blocks in the file
   (including indirect blocks). */
#ifndef HAVE_ST_BLOCKS
# define ST_BLKSIZE(statbuf) DEV_BSIZE
# if defined(_POSIX_SOURCE) || !defined(BSIZE) /* fileblocks.c uses BSIZE.  */
#  define ST_NBLOCKS(statbuf) (((statbuf).st_size + 512 - 1) / 512)
# else /* !_POSIX_SOURCE && BSIZE */
#  define ST_NBLOCKS(statbuf) (st_blocks ((statbuf).st_size))
# endif /* !_POSIX_SOURCE && BSIZE */
#else /* HAVE_ST_BLOCKS */
/* Some systems, like Sequents, return st_blksize of 0 on pipes. */
# define ST_BLKSIZE(statbuf) ((statbuf).st_blksize > 0 \
			       ? (statbuf).st_blksize : DEV_BSIZE)
# if defined(hpux) || defined(__hpux__) || defined(__hpux)
/* HP-UX counts st_blocks in 1024-byte units.
   This loses when mixing HP-UX and BSD filesystems with NFS.  */
#  define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 2)
# else /* !hpux */
#  if defined(_AIX) && defined(_I386)
/* AIX PS/2 counts st_blocks in 4K units.  */
#   define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 8)
#  else /* not AIX PS/2 */
#   if defined(_CRAY)
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * ST_BLKSIZE(statbuf)/512)
#   else /* not AIX PS/2 nor CRAY */
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks)
#   endif /* not _CRAY */
#  endif /* not AIX PS/2 */
# endif /* !hpux */
#endif /* HAVE_ST_BLOCKS */

/* Convert B 512-byte blocks to kilobytes if K is nonzero,
   otherwise return it unchanged. */
#define convert_blocks(b, k) ((k) ? ((b) + 1) / 2 : (b))

#ifndef RETSIGTYPE
#define RETSIGTYPE void
#endif

#ifdef __GNUC__
# undef alloca
# define alloca __builtin_alloca
#else
# ifdef HAVE_ALLOCA_H
#  include <alloca.h>
# else
#  ifndef _AIX
/* AIX alloca decl has to be the first thing in the file, bletch! */
char *alloca ();
#  endif
# endif
#endif

#include <ctype.h>

/* Jim Meyering writes:

   "... Some ctype macros are valid only for character codes that
   isascii says are ASCII (SGI's IRIX-4.0.5 is one such system --when
   using /bin/cc or gcc but without giving an ansi option).  So, all
   ctype uses should be through macros like ISPRINT...  If
   STDC_HEADERS is defined, then autoconf has verified that the ctype
   macros don't need to be guarded with references to isascii. ...
   Defining isascii to 1 should let any compiler worth its salt
   eliminate the && through constant folding."

   Bruno Haible adds:

   "... Furthermore, isupper(c) etc. have an undefined result if c is
   outside the range -1 <= c <= 255. One is tempted to write isupper(c)
   with c being of type `char', but this is wrong if c is an 8-bit
   character >= 128 which gets sign-extended to a negative value.
   The macro ISUPPER protects against this as well."  */

#if defined (STDC_HEADERS) || (!defined (isascii) && !defined (HAVE_ISASCII))
#define IN_CTYPE_DOMAIN(c) 1
#else
#define IN_CTYPE_DOMAIN(c) isascii(c)
#endif

#ifdef isblank
#define ISBLANK(c) (IN_CTYPE_DOMAIN (c) && isblank (c))
#else
#define ISBLANK(c) ((c) == ' ' || (c) == '\t')
#endif
#ifdef isgraph
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isgraph (c))
#else
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isprint (c) && !isspace (c))
#endif

#define ISPRINT(c) (IN_CTYPE_DOMAIN (c) && isprint (c))
#define ISALNUM(c) (IN_CTYPE_DOMAIN (c) && isalnum (c))
#define ISALPHA(c) (IN_CTYPE_DOMAIN (c) && isalpha (c))
#define ISCNTRL(c) (IN_CTYPE_DOMAIN (c) && iscntrl (c))
#define ISLOWER(c) (IN_CTYPE_DOMAIN (c) && islower (c))
#define ISPUNCT(c) (IN_CTYPE_DOMAIN (c) && ispunct (c))
#define ISSPACE(c) (IN_CTYPE_DOMAIN (c) && isspace (c))
#define ISUPPER(c) (IN_CTYPE_DOMAIN (c) && isupper (c))
#define ISXDIGIT(c) (IN_CTYPE_DOMAIN (c) && isxdigit (c))
#define ISDIGIT_LOCALE(c) (IN_CTYPE_DOMAIN (c) && isdigit (c))

/* ISDIGIT differs from ISDIGIT_LOCALE, as follows:
   - Its arg may be any int or unsigned int; it need not be an unsigned char.
   - It's guaranteed to evaluate its argument exactly once.
   - It's typically faster.
   Posix 1003.2-1992 section 2.5.2.1 page 50 lines 1556-1558 says that
   only '0' through '9' are digits.  Prefer ISDIGIT to ISDIGIT_LOCALE unless
   it's important to use the locale's definition of `digit' even when the
   host does not conform to Posix.  */
#define ISDIGIT(c) ((unsigned) (c) - '0' <= 9)

#ifndef __P
#if defined (__GNUC__) || (defined (__STDC__) && __STDC__)
#define __P(args) args
#else
#define __P(args) ()
#endif  /* GCC.  */
#endif  /* Not __P.  */

/* Take care of NLS matters.  */

#if HAVE_LOCALE_H
# include <locale.h>
#endif
#if !HAVE_SETLOCALE
# define setlocale(Category, Locale) /* empty */
#endif

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# undef bindtextdomain
# define bindtextdomain(Domain, Directory) /* empty */
# undef textdomain
# define textdomain(Domain) /* empty */
# define _(Text) Text
#endif

#define STREQ(a,b) (strcmp((a), (b)) == 0)


src/DOTparser.h
/* 
  DOTparser.h
  Description: Simple dot language parser implemented using regular
  expressions
  ----------------------------------------------------------------------------- 

  Started on  <Fri Jul  1 09:08:22 2016 Carlos Linares Lopez>
  Last update <jueves, 21 julio 2016 10:30:14 Carlos Linares Lopez (clinares)>
  -----------------------------------------------------------------------------

  $Id::                                                                      $
  $Date::                                                                    $
  $Revision::                                                                $
  -----------------------------------------------------------------------------

  Made by Carlos Linares Lopez
  Login   <clinares@atlas>
*/

#ifndef   	PARSER_H_
# define   	PARSER_H_

#include "DOTdefs.h"

#include <fstream>
#include <iostream>
#include <map>
#include <regex>
#include <string>
#include <utility>
#include <vector>

namespace dot {

  using namespace std;

  class parser {

  public:

    // Default constructor
    parser ()
      : _filename {""}
    { }
    
    // Explicit constructor
    parser (string filename)
      : _filename {filename}
    { }

    // get/set accessors
    string get_type () const
    { return _type; }
    string get_name () const
    { return _name; }

    // show a void line
    void show_void (const string& line, bool verbose);
    
    // get all vertices of the graph
    vector<string> get_vertices () const;
    
    // get all nodes that are reachable from a given node
    vector<string> get_neighbours (const string name);
    
    // get all the attributes of the specified vertex
    vector<string> get_vertex_attributes (const string name);
  
    // return the value of an attribute defined for a specific vertex
    string get_vertex_attribute (string name, string attrname);

    // get all the attributes of a specific edge qualified by its
    // (origin,target) names
    vector<string> get_edge_attributes (const string origin, const string target);

    // return the value of the given attribute defined for the vertex qualified
    // by (origin,target) names
    string get_edge_attribute (const string origin, const string target, const string attr);

    // set the filename to parse
    void set_filename (string filename)
    { _filename = filename; }
    
    // Sets the verbose level to true by default
    void set_verbose (bool value = true)
    { _verbose = value; }
    
    // parse the file given in the construction of this instance. It returns
    // true if the file could be successfully parse and false otherwise.
    bool parse ();    
    
  private:

    // return in contents all the contents of the file in _filename where
    // newlines have been removed
    bool _read_file (string& contents);
    
    // the following methods parse different types according to the given
    // regexp. The second and third methods return in value the matching of the
    // given regexp.
    bool _parse_void (string& content, const string& regexp);
    bool _parse_string (string& content, const string& regexp, string& value);
    bool _parse_unsigned_int (string& content, const string& regexp, 
			      unsigned int& value);

    // skips the section of contents if it contains comments. It returns the
    // contents that result after stripping the leading comments. It honors both
    // C and C++ comments (as specified in the dot language)
    void _parse_comments (string& contents);

    // parse the given contents looking for the graph type. It returns true if
    // it could find the type and false otherwise. In case the graph type is
    // successfully determined, it is returned in type.
    bool _parse_graph_type (string& contents, string& type);

    // parse the given contents looking for the graph name. It returns true if
    // it could find the graph name and false otherwise.  In case the graph name
    // is successfully determined, it is returned in name.
    bool _parse_graph_name (string& contents, string& name);

    // parse the given contents looking for the beginning of a block. It returns
    // true if it is found and false otherwise.
    bool _parse_block_begin (string& contents);

    // parse the given contents looking for the name of a vertex. It returns
    // true if it is found and false otherwise.  In case the vertex name is
    // successfully determined, it is returned in name.
    bool _parse_vertex_name (string& contents, string& name);

    // parse the given contents looking for the type of an edge.  It returns
    // true if it is found and false otherwise.  In case the edge type is
    // successfully determined, it is returned in edge_type.
    bool _parse_edge_type (string& contents, string& edge_type);

    // parse the given contents looking for the name of a target vertex. It then
    // uses the name of the origin vertex and the type of the edge type (that
    // should have been previously processed) to update the adjacency map
    // describing the graph. It returns true if a target could be successfully
    // processed, and false otherwise.
    bool _parse_target_name (string& contents, string& name,
			     string orig_name, string edge_type);

    // parse an attributes section. The attributes read are return as a map that
    // stores for every attribute its value as a string. It returns true if any
    // attributes were found and false otherwise.
    bool _parse_attributes (string& contents, map<string, string>& dict);
    
  private:

    // INVARIANTS: every dot parser consists of a filename which is stored in a
    // private data member. The parser is responsible to process the contents of
    // this file sequentially discanding those parts that are successfully
    // processed
    string _filename;           // filename with the dot specification to parse

    // INVARIANTS: each graph has a type (either directed or undirected) and a
    // name (which can be the empty string). They are stored in dedicated data
    // members
    string _type;                  // graph type, either directed or undirected
    string _name;              // graph name which can be also the empty string

    // INVARIANTS: graphs are represented like adjacency maps which are indexed
    // by the vertex name given in the dot file. Each entry of the map consists
    // then of a vector of adjacent vertices
    map<string, vector<string>> _graph;
    
    // INVARIANTS: each graph is made of vertices and edges which can be
    // characterized with attributes. These are stored in a couple of
    // maps. _vertex stores for every vertex name a map of attributes of the
    // form "name"->"value". _edge stores for every vertex name a map which is
    // indexed with the target vertex name and a map of attribute names and
    // values. In both cases, though values can be casted into different types,
    // they are internally stored as strings.
    map<string,map<string,string>> _vertex;
    map<string,map<string,map<string, string>>> _edge;

    // Additionally, the user can explicitly request verbose output while
    // parsing the dot specification file. It is disabled by default
    bool _verbose;                     // whether the user wants verbose output
    
  };  // class parser
  
} // namespace dot

#endif 	    /* !PARSER_H_ */


/* Local Variables: */
/* mode:c++ */
/* fill-column:80 */
/* End: */


parser/xmalloc.c
/* xmalloc.c -- malloc with out of memory checking
   Copyright (C) 1990, 91, 92, 93, 94, 95, 96, 99 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#if __STDC__
# define VOID void
#else
# define VOID char
#endif

#include <stdio.h>		/* for stderr */

#if STDC_HEADERS

#include <sys/types.h>
#include <string.h>		/* for strlen etc. */
#include <stdlib.h>

#else  /* !STDC_HEADERS */

extern size_t strlen ();
extern char *strcpy ();

VOID *calloc ();
VOID *malloc ();
VOID *realloc ();
void free ();
#endif

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# define textdomain(Domain)
# define _(Text) Text
#endif

#include <error.h>

/* Prototypes for functions defined here.  */
#if defined (__STDC__) && __STDC__
static VOID *fixup_null_alloc (size_t n);
VOID *xmalloc (size_t n);
VOID *xcalloc (size_t n, size_t s);
VOID *xrealloc (VOID *p, size_t n);
char *xstrdup (char *p);
#endif


static VOID *
fixup_null_alloc (n)
     size_t n;
{
  VOID *p;

  p = 0;
  if (n == 0)
    p = malloc ((size_t) 1);
  if (p == 0)
    {
      /* possible revisions: release some memory and re-try, print
	 more information (e.g. line number of input file) */
      fprintf(stderr, _("parser: Memory exhausted"));
      exit(1);
    }
  return p;
}

/* Allocate N bytes of memory dynamically, with error checking.  */

VOID *
xmalloc (n)
     size_t n;
{
  VOID *p;

  p = malloc (n);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Allocate memory for N elements of S bytes, with error checking.  */

VOID *
xcalloc (n, s)
     size_t n, s;
{
  VOID *p;

  p = calloc (n, s);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Change the size of an allocated block of memory P to N bytes,
   with error checking.
   If P is NULL, run xmalloc.  */

VOID *
xrealloc (p, n)
     VOID *p;
     size_t n;
{
  if (p == 0)
    return xmalloc (n);
  p = realloc (p, n);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Make a copy of a string in a newly allocated block of memory. */

char *
xstrdup (str)
     char *str;
{
  VOID *p;

  p = xmalloc (strlen (str) + 1);
  strcpy (p, str);
  return p;
}


parser/parser.cc
/* 
   parser - DOT parser built with the DOT parser library

   Copyright (C) 2016 Carlos Linares Lopez

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

*/

#include <src/DOTparser.h>

#include <iostream>
#include <string>

#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "system.h"

using namespace std;

#define EXIT_NICE 0
#define EXIT_FAILURE 1

extern "C" {
char *xstrdup (char *p);
}

// *** Globals
char *program_name;

// *** Option flags and variables
static struct option const long_options[] =
{
  {"file", required_argument, 0, 'f'},
  {"show-attributes", no_argument, 0, 's'},
  {"verbose", no_argument, 0, 'v'},
  {"help", no_argument, 0, 'h'},
  {"version", no_argument, 0, 'V'},
  {NULL, 0, NULL, 0}
};

// ** Prototypes

// Set all the option flags according to the switches specified.  Return the
// index of the first non-option argument.
static int decode_switches (int argc, char **argv, string& file, bool& show_attributes,
			      bool& want_verbose);

static void usage (int status);

int
main (int argc, char **argv)
{
  
  string file;              // file with the dot specification of this instance
  bool show_attributes;                // are attributes requested to be shown?
  bool want_verbose;        // whether verbose output has been requested or not

  string name;                                                    // graph name
  string type;                                                    // graph type
  
  program_name = argv[0];
  decode_switches (argc, argv, file, show_attributes, want_verbose);

  // *** parameter checking
  if (!file.size ()) {

    cerr << "\n Please, provide the file with the dot specification with --file" << endl;
    cerr << " See " << program_name << " --help for more details" << endl << endl;
    exit(EXIT_FAILURE);
  }
  
  // *** do the work

  // create an objet of a dot parser
  dot::parser parser (file);

  // set the verbosity level
  parser.set_verbose (want_verbose);

  // and parse its contents
  if (want_verbose)
    cout << endl;
  if (!parser.parse ())
    return EXIT_FAILURE;

  // show admin info of the graph
  cout << endl;
  cout << " Type: " << parser.get_type () << endl;
  cout << " Name: " << parser.get_name () << endl << endl;

  // in case the attributes have been required
  if (show_attributes) {

    // *** vertex attributes

    // get all vertices of the graph
    vector<string> vertices = parser.get_vertices ();
    if (vertices.size ()) {
      for (auto ivertex = vertices.begin () ; ivertex != vertices.end () ; ++ivertex) {

	// now, get all attributes for this specific vertex
	vector<string> attrs = parser.get_vertex_attributes (*ivertex);
	if (attrs.size ()) {
	  cout << " Vertex " << *ivertex << ":" << endl;
	  for (auto iattr = attrs.begin () ; iattr != attrs.end () ; ++iattr)

	    cout << "\t " << *iattr << ": " << parser.get_vertex_attribute (*ivertex, *iattr) << endl;
	  cout << endl;
	}
      }
    }
    
    // *** edge attributes

    // for every pair of vertices which are connected through an edge
    for (auto ivertex = vertices.begin () ;  ivertex != vertices.end () ; ++ivertex) {

      // get all nodes that are immediately reachable from this one
      vector<string> neighbours = parser.get_neighbours (*ivertex);
      for (auto jvertex = neighbours.begin () ; jvertex != neighbours.end () ; ++jvertex ) {

	// get all attributes of this edge
	vector<string> attrs = parser.get_edge_attributes (*ivertex, *jvertex);
	if (attrs.size ()) {

	  cout << " Edge " << *ivertex << " -> " << *jvertex << ":" << endl;

	  // and show all attributes of this edge
	  for (auto iattr = attrs.begin () ; iattr != attrs.end () ; ++iattr ) {
	    cout << "\t " << *iattr << ": " << parser.get_edge_attribute (*ivertex, *jvertex, *iattr) << endl;
	  }
	  cout << endl;
	}
      }
    }
  }

  // and exit
  return EXIT_NICE;
}

// Set all the option flags according to the switches specified.  Return the
// index of the first non-option argument.
static int
decode_switches (int argc, char **argv, string& file, bool& show_attributes,
		 bool& want_verbose)
{
  int c;

  // default values
  file = "";
  show_attributes = false;
  want_verbose = false;
  
  while ((c = getopt_long (argc, argv, 
			   "f"  /* file */
			   "s"  /* show-attributes */
			   "v"	/* verbose */
			   "h"	/* help */
			   "V",	/* version */
			   long_options, (int *) 0)) != EOF) {
    switch (c) {
    case 'f':		/* --file */
      file = optarg;
      break;
    case 's':		/* --show-attributes */
      show_attributes = 1;
      break;
    case 'v':		/* --verbose */
      want_verbose = 1;
      break;
    case 'V':
      printf ("DOT parser %s\n", VERSION);
      exit (0);
      
    case 'h':
      usage (0);
      
    default:
      usage (EXIT_FAILURE);
    }
  }
  
  return optind;
}

static void
usage (int status)
{
  cout << endl << " " << program_name << " parses dot files" << endl << endl;
  cout << " Usage: " << program_name << " [OPTION]" << endl << endl;
  cout << "\
 Options:\n\
      -f, --file                 file with the dot specification of the graph\n\
                                 see the README.md with information on the subset\n\
                                 of the dot language used\n\
      -s, --show-attributes      if given, all vertex and edge attributes are printed\n\
                                 on the console\n\
\n\
      --verbose                  print verbose information\n\
      -h, --help                 display this help and exit\n\
      -V, --version              output version information and exit\n\
\n";
  exit (status);
}

/* Local Variables: */
/* mode:c++ */
/* fill-column:80 */
/* End: */
